rumqttc/src/v5/client/asyncclient.rs

```diff
@@ -0,0 +1,314 @@
+        let pkid = self.increment_pkid();
+        publish.pkid = pkid;
+        self.send_async_and_notify(Request::Publish(publish)).await?;
+        Ok(pkid)
+    }
+
+    /// Sends a MQTT Subscribe to the eventloop
+    pub async fn subscribe<S: Into<String>>(&self, topic: S, qos: QoS) -> Result<(), ClientError> {
+        let subscribe = Subscribe::new(topic.into(), qos);
+        self.send_async_and_notify(Request::Subscribe(subscribe))
+            .await
+    }
+
+    /// Sends a MQTT Subscribe to the eventloop
+    pub fn try_subscribe<S: Into<String>>(&self, topic: S, qos: QoS) -> Result<(), ClientError> {
+        let subscribe = Subscribe::new(topic.into(), qos);
+        self.try_send_and_notify(Request::Subscribe(subscribe))
+    }
+
+    /// Sends a MQTT Subscribe for multiple topics to the eventloop
+    pub async fn subscribe_many<T>(&self, topics: T) -> Result<(), ClientError>
+    where
+        T: IntoIterator<Item = SubscribeFilter>,
+    {
+        let subscribe = Subscribe::new_many(topics);
+        self.send_async_and_notify(Request::Subscribe(subscribe))
+            .await
+    }
+
+    /// Sends a MQTT Subscribe for multiple topics to the eventloop
+    pub fn try_subscribe_many<T>(&self, topics: T) -> Result<(), ClientError>
+    where
+        T: IntoIterator<Item = SubscribeFilter>,
+    {
+        let subscribe = Subscribe::new_many(topics);
+        self.try_send_and_notify(Request::Subscribe(subscribe))
+    }
+
+    /// Sends a MQTT Unsubscribe to the eventloop
+    pub async fn unsubscribe<S: Into<String>>(&self, topic: S) -> Result<(), ClientError> {
+        let unsubscribe = Unsubscribe::new(topic.into());
+        self.send_async_and_notify(Request::Unsubscribe(unsubscribe))
+            .await
+    }
+
+    /// Sends a MQTT Unsubscribe to the eventloop
+    pub fn try_unsubscribe<S: Into<String>>(&self, topic: S) -> Result<(), ClientError> {
+        let unsubscribe = Unsubscribe::new(topic.into());
+        self.try_send_and_notify(Request::Unsubscribe(unsubscribe))
+    }
+
+    /// Sends a MQTT disconnect to the eventloop
+    pub async fn disconnect(&self) -> Result<(), ClientError> {
+        self.send_async_and_notify(Request::Disconnect).await
+    }
+
+    /// Sends a MQTT disconnect to the eventloop
+    pub fn try_disconnect(&self) -> Result<(), ClientError> {
+        self.try_send_and_notify(Request::Disconnect)
+    }
+
+    /// Stops the eventloop right away
+    pub async fn cancel(&self) -> Result<(), ClientError> {
+        self.cancel_tx
+            .send_async(())
+            .await
+            .map_err(ClientError::Cancel)
+    }
+
+    async fn send_async_and_notify(&self, request: Request) -> Result<(), ClientError> {
+        {
+            let mut request_buf = self.request_buf.lock().unwrap();
+            if request_buf.len() == self.request_buf_capacity {
+                return Err(ClientError::RequestsFull);
+            }
+            request_buf.push_back(request);
```

Comment:
Create a struct with `outgoing_buf` and `pkid` and wrap it with `Arc Mutex` instead of maintaining `pkid` and an atomic.

Increment `pkid` here after add data to buffer